trio/tests/test_wait_for_single_object.py

```diff
@@ -0,0 +1,168 @@
+import time
+from threading import Thread
+
+import pytest
+
+on_windows = (os.name == "nt")
+# Mark all the tests in this file as being windows-only
+pytestmark = pytest.mark.skipif(not on_windows, reason="windows only")
+
+from .._core.tests.tutil import slow
+from .. import _core
+from .. import _timeouts
+if on_windows:
+    from .._core._windows_cffi import ffi, kernel32
+    from .._wait_for_single_object import WaitForSingleObject, WaitForMultipleObjects_sync
+
+
+async def test_WaitForMultipleObjects_sync():
+    # This does a series of tests where we set/close the handle before
+    # initiating the waiting for it.
+    #
+    # Note that closing the handle (not signaling) will cause the
+    # *initiation* of a wait to return immediately. But closing a handle
+    # that is already being waited on will not stop whatever is waiting
+    # for it.
+
+    # One handle
+    handle1 = kernel32.CreateEventA(ffi.NULL, True, False, ffi.NULL)
+    kernel32.SetEvent(handle1)
+    WaitForMultipleObjects_sync(handle1)
+    kernel32.CloseHandle(handle1)
+
+    # Two handles, signal first
+    handle1 = kernel32.CreateEventA(ffi.NULL, True, False, ffi.NULL)
+    handle2 = kernel32.CreateEventA(ffi.NULL, True, False, ffi.NULL)
+    kernel32.SetEvent(handle1)
+    WaitForMultipleObjects_sync(handle1, handle2)
+    kernel32.CloseHandle(handle1)
+    kernel32.CloseHandle(handle2)
+
+    # Two handles, signal second
+    handle1 = kernel32.CreateEventA(ffi.NULL, True, False, ffi.NULL)
+    handle2 = kernel32.CreateEventA(ffi.NULL, True, False, ffi.NULL)
+    kernel32.SetEvent(handle2)
+    WaitForMultipleObjects_sync(handle1, handle2)
+    kernel32.CloseHandle(handle1)
+    kernel32.CloseHandle(handle2)
+
+    # Two handles, close first
+    handle1 = kernel32.CreateEventA(ffi.NULL, True, False, ffi.NULL)
+    handle2 = kernel32.CreateEventA(ffi.NULL, True, False, ffi.NULL)
+    kernel32.CloseHandle(handle1)
+    WaitForMultipleObjects_sync(handle1, handle2)
+    kernel32.CloseHandle(handle2)
+
+    # Two handles, close second
+    handle1 = kernel32.CreateEventA(ffi.NULL, True, False, ffi.NULL)
+    handle2 = kernel32.CreateEventA(ffi.NULL, True, False, ffi.NULL)
+    kernel32.CloseHandle(handle2)
+    WaitForMultipleObjects_sync(handle1, handle2)
+    kernel32.CloseHandle(handle1)
+
+
+@slow
+async def test_WaitForMultipleObjects_sync_slow():
+    # This does a series of test in which the main thread sync-waits for
+    # handles, while we spawn a thread to set the handles after a short while.
+
+    TIMEOUT = 0.3
+
+    def signal_soon_sync(handle):
+        time.sleep(TIMEOUT)
+        kernel32.SetEvent(handle)
+
+    # One handle
+    handle1 = kernel32.CreateEventA(ffi.NULL, True, False, ffi.NULL)
+    t0 = _core.current_time()
+    Thread(target=signal_soon_sync, args=(handle1,)).start()
+    WaitForMultipleObjects_sync(handle1)
+    t1 = _core.current_time()
+    assert TIMEOUT <= (t1 - t0) < 1.2 * TIMEOUT
+    kernel32.CloseHandle(handle1)
```

Comment:
So... this is weird. If you look at the Appveyor results, the `sync_slow` test failed on a bunch of trials, though the other tests passed, including the other `_slow` test below. And the reason they failed is that `t1 - t0` ended up being like 0.297, which is less than 0.3. Which seems like it should be impossible.

I think that's going on is that on Windows, `time.sleep` actually uses a slightly different clock than trio does. The Windows clock APIs are really complicated, but trio currently uses `time.monotonic`, which is not quite the same as other clocks. If you use `trio.sleep`, that's always fully in sync with `trio.current_time`, so the assertion should pass – and that's what we see in the other slow test below. But if you use `time.sleep`, then apparently there can be a bit of skew.

So, I think you should rewrite this test to use `await trio.sleep`, like the other one below. Fortunately this fits in nicely with my other comment about not using the raw threading API :-).

Also, I'd switch all the timeouts checks to use `< 2 * TIMEOUT`, based on what we've ended up doing in `test_run.py` and `test_ki.py`. (We used to use smaller values and then would get false failures when some CI server was running slow, and eventually ended up bumping the values up to that.)

Reply:
(I think I fixed it (yet unpushed) by checking `time.monotonic`. Am happy to change to make it more trio-ish, though not sure how that would work)

Reply:
Right, you can't use `trio.sleep()` inside a sync function. But you can move `trio.sleep()` into the main thread, and put the `WaitForMultipleObjects_sync` into a thread :-).

Reply:
Right :)